test/resources/knativeresources.go

```diff
@@ -91,7 +91,24 @@ func IsKnativeDeploymentReady(dpList *v1.DeploymentList, expectedDeployments []s
 			// Currently, the network ingress resource is still specified together with the knative serving.
 			// It is possible that network ingress resource is not using the same version as knative serving.
 			// This is the reason why we skip the version checking for network ingress resource.
-			if val == fmt.Sprintf("v%s", version) || version == common.LATEST_VERSION || key == "networking.knative.dev/ingress-provider" {
+			statusCheck := false
+			if val == fmt.Sprintf("v%s", version) && version != common.LATEST_VERSION {
+				statusCheck = true
+			}
+
+			if key == "networking.knative.dev/ingress-provider" {
+				statusCheck = true
+			}
+
+			if version == common.LATEST_VERSION && version == existingVersion {
+				statusCheck = true
+			}
+
+			if version == common.LATEST_VERSION && version != existingVersion && (key == "serving.knative.dev/release" || key == "eventing.knative.dev/release") && val != fmt.Sprintf("v%s", existingVersion) {
```

Comment:
I don't quite understand the extra logic here -- might be worth a comment, or rewriting as:

```golang
if version == common.LATEST_VERSION && version != existingVersion {
  if (key == "serving.knative.dev/release" || key = "eventing.knative.dev/release") && val != fmt.Sprintf("v%s", existingVersion) {
    statusCheck = true
  }
}
```

There's such a variety of cases here, I'm struggling to figure out what the common axes are. Reading a bit further, it looks like we know a `version` and `existingVersion` from some prior knowledge, and are checking labels for matches. It feels like we should have an early-exit on the loop for labels where the key doesn't match, e.g.

```golang
for k, v := range d.GetObjectMeta().GetLabels() {
  if key == "networking.knative.dev/ingress-provider" {
    break
  }

  if key == "serving.knative.dev/release" || key == "serving.knative.dev/release" {
    // If versions don't match
    return false
  }
}
return isReady(d.Status)  // At least check the Deployment conditions, but there might be other "ready" statuses that need checking.

See https://carvel.dev/kapp/docs/latest/apply-waiting/#overview for some other examples of "waiting" things (`batch/v1/Job` in particular).

Reply:
In particular, I'm concerned about:

```golang
if val == fmt.Sprintf("v%s", version) && version != common.LATEST_VERSION
```

possibly being able to pick up unexpected label keys.

Reply:
I will add comments as many as possible, but I would like to keep the existing logic.